sdk/core/System.BinaryData/src/ReadOnlyMemoryStream.cs

```diff
@@ -8,7 +8,7 @@
 using System.Threading;
 using System.Threading.Tasks;
 
-namespace Azure.Core
+namespace System
 {
     /// <summary>Provides a <see cref="Stream"/> for the contents of a <see cref="ReadOnlyMemory{Byte}"/>.</summary>
     internal sealed class ReadOnlyMemoryStream : Stream
```

Comment:
There are a couple of issues that I see with this type; I don't know what we want to do about it.

* It doesn't have a Dispose implementation, so it will keep allowing reads after a Dispose
* If Position gets set to a value in excess of Length it succeeds (so does MemoryStream) and then a read returns 0 (MemoryStream throws a "beyond EOF" exception)
* The result of `ToStream` still shows mutation:

```C#
byte[] buf = { 1, 2, 3, 4, 5 };
BinaryData data = new BinaryData(buf);
buf[0] = 2;
Stream stream = data.ToStream();
buf[0] = 3;

byte first = stream.ReadByte();
```

<details>
  <summary>What value does `first` have?</summary>

   3
</details>

The mutation-visible aspect is an implementation detail that leaks out as a consequence of the ctors not doing defensive copies.  Maybe it's OK.  The original implementation of ReadOnlyMemoryStream is using a buffer wholly owned by the issuing object, so it doesn't have the mutation concern.

Reply:
As you mentioned, since BinaryData is mutable when constructed from bytes, the fact that a stream created from it is also mutable seems reasonable. 

As far as adding a dispose implementation to prevent reading after dispose, and the Position validation, do we think it is worth aligning with MemoryStream here?

Reply:
> the fact that a stream created from it is also mutable seems reasonable.

I don't really agree.   Aside from ToMemory (which should have been AsMemory, but, compromises), and ToStream (which we're discussing), the To methods always produce something which is detached from the data.  Part of the To vs As debate was that To is often/usually a clean object (no side effects bleed through) and As is a linked object (side effects do bleed through).

Since I can understand the rationale for wrapping the input bytes, I think my preference would be to make the object track whether or not it "owns" the buffer, and for a wrapped buffer ToStream should pass a copy, so it's side effect free.

> As far as adding a dispose implementation to prevent reading after dispose

I think it should be added.  Dispose clearing the buffer reference might make the difference between whether or not the underlying data is available for garbage collection (if the Stream reference outlives the data's usefulness).

> As far as ... the Position validation, 

It looks like FileStream (at least the one produced by File.OpenRead) allows seeking to beyond the end and then just returns EOF (same as the implementation here).  So since MemoryStream is the odd one out, I'm OK with leaving the Position validation alone.

Reply:
I would look at ToStream the same way as ToMemory. We would've called it AsStream if not for the consistency issue.

Reply:
> To methods always produce something which is detached from the data.

"hello jeremy".ToString() does not produce a detached object. I know it's an extreme case but there are others. In other words, I don't think there are any expectations of detaching. I think we just need to ensure that the stream is readonly and does not allow callers to see bytes before and after the slice (as pointed out in another comment)

Reply:
My "always" was scoped to the methods on this type.  BinaryData::ToString() or BinaryData::ToObjectFromJson&lt;T> produce outputs that do not mutate as a result of writing back to the original buffer.

For a symmetry argument, since `FromStream` eagerly consumes the data from the stream (once the method terminates further writes to the stream have no effect on the BinaryData), it's weird (to me) if ToStream produces a seekable stream where seeking backwards and re-reading could produce different results.  (Two different calls to ToStream producing streams that have different answers is OK, it matches what would happen for a JSON deserialization, then mutation, then a second deserialization)

Reply:
>  if ToStream produces a seekable stream where seeking backwards and re-reading could produce different results.

What's the code that would do this? BinaryData is a "readable" type, i.e. the code that constructs it can mutable the data, but not the code that uses the type. 

Besides, what you point out is an intrinsic property/quality of non-deep-immutable types. We have thousands of these in .NET. 

Overall, I feel like the main practical result of this change would be to make the type very inefficient (copy large buffers).